source/android/mobile/src/main/java/io/adaptivecards/adaptivecardssample/RemoteClientConnection.java

```diff
@@ -0,0 +1,501 @@
+        this.m_observer = observer;
+    }
+
+    public interface Observer
+    {
+        void onConnecting(String status);
+
+        void onConnectFailed(String errorMessage);
+
+        void onCardPayload(String cardPayload);
+
+        void onHostConfigPayload(String hostConfigPayload);
+
+        void onStateChanged(State state);
+    }
+
+    public void connect(String hostId)
+    {
+        m_hostId = hostId;
+        RequestQueue queue = Volley.newRequestQueue(m_context);
+        String url = "https://signalingserver.azurewebsites.net/api/subscriptions/" + hostId;
+
+        // Request the offer info
+        JsonObjectRequest stringRequest = new JsonObjectRequest(Request.Method.GET, url, null, new Response.Listener<JSONObject>()
+        {
+            @Override
+            public void onResponse(JSONObject response)
+            {
+                connectToHost(response);
+            }
+        }, new Response.ErrorListener()
+        {
+            @Override
+            public void onErrorResponse(VolleyError error)
+            {
+                m_observer.onConnectFailed("Is your internet working? Failed to request offer info.");
+                changeStateToClosed();
+            }
+        });
+
+        // Add the request to the queue
+        m_observer.onConnecting("Requesting offer info...");
+        queue.add(stringRequest);
+    }
+
+    public void disconnect()
+    {
+        if (m_conn != null)
+        {
+            m_conn.dispose();
+        }
+
+        m_observer = null;
+    }
+
+    private void connectToHost(final JSONObject offerAndCandidates)
+    {
+        m_observer.onConnecting("Accepting offer...");
+
+        // Code from https://youtu.be/HS1eKPL4f1o
+        PeerConnectionFactory.InitializationOptions options = PeerConnectionFactory.InitializationOptions.builder(m_context).createInitializationOptions();
+        PeerConnectionFactory.initialize(options);
+
+        PeerConnectionFactory factory = PeerConnectionFactory.builder().createPeerConnectionFactory();
+
+        ArrayList<PeerConnection.IceServer> iceServers = new ArrayList<>();
+        iceServers.add(PeerConnection.IceServer.builder("stun:ws-turn3.xirsys.com").createIceServer());
+        String[] turnServers = new String[]
+            {
+                "turn:ws-turn3.xirsys.com:80?transport=udp",
+                "turn:ws-turn3.xirsys.com:3478?transport=udp",
+                "turn:ws-turn3.xirsys.com:80?transport=tcp",
+                "turn:ws-turn3.xirsys.com:3478?transport=tcp",
+                "turns:ws-turn3.xirsys.com:443?transport=tcp",
+                "turns:ws-turn3.xirsys.com:5349?transport=tcp"
+            };
+        for (String turnServer : turnServers)
+        {
+            iceServers.add(PeerConnection.IceServer.builder(turnServer).setUsername("e-HcX5lqoTEzSQetY5biyT3YzM45GIl3HK4FQuo3y73xvsLAmts_gF8PylqkwBR6AAAAAFyoMKphbGVhZGVy").setPassword("d97cbeee-5827-11e9-bef6-069f0817bf63").createIceServer());
```

Comment:
>nServer).setUsername("e-HcX5lqoTEzSQetY5biyT3YzM45GIl3HK4FQuo3y73xvsLAmts_gF8PylqkwBR6AAAAAFyoMKphbGVhZGVy").setPassword("d97cbeee-5827-11e9-bef6-069f0817bf63") [](start = 63, length = 160)

is it ok to have this credentials like this?

Reply:
It's for a free TURN server, not too concerned... we could always switch it to obtain the passwords somehow else, but inevitably these are always a bit vulnerable to being taken and used. Worst case is that someone uses it and consumes all the bandwidth on the TURN server, which would simply result in our QR sharing service being slow/unusable, but wouldn't charge us any money since it's free. Most connections don't use TURN anyways though. We can re-visit if we run into problems!